### PR TITLE
qbot/gerrit_notif: add a qbot which handles Gerrit notif

### PR DESCRIPTION
**Pull request by yannKagan**
Forward gerrit notifications to slack using qbot.

### CODE REVIEW

File: scripts/event-handler.coffee [hunk 1]
```diff
@@ -77,3 +90,31 @@ module.exports = (robot) ->
     # Send notification to watchers
     for idx,w of details.watchers
       notify_user w.login
+
+  robot.on 'comment_added', (details) ->
```
Comment (vthib) [inline scripts/event-handler.coffee]: I'd rather have a event name that includes 'gerrit', 'gerrit-notif' to mirror what is done for redmine

File: scripts/event-handler.coffee [hunk 2]
```diff
@@ -77,3 +90,31 @@ module.exports = (robot) ->
...
+        {
+          title: details.subject
+          title_link: details.change_url
+          text: details.comment
+          color: get_color_from_comment details.comment
+        }
+      ]
+      as_user: true
+    }
+    notified_gerrit = {}
```
Comment (vthib) [inline scripts/event-handler.coffee]: useless

File: scripts/event-handler.coffee [hunk 3]
```diff
@@ -77,3 +90,31 @@ module.exports = (robot) ->
...
+      if details.emitter
+        text = details.author.replace /<.*>/, "@#{details.emitter}"
+      else
+        text = details.author
+
+      robot.emit 'user-send', login, 'gerrit', text, msg
+      notified_gerrit[login] = true
+
+    # Send notification to owner of patch
+    if details.author != details.change_owner
```
Comment (vthib) [inline scripts/event-handler.coffee]: doing this before building 'msg' would be better

File: scripts/gerrit-handler.coffee [hunk 4]
```diff
@@ -0,0 +1,74 @@
+# Handle Gerrit webhook notifications
```
Comment (vthib) [inline scripts/gerrit-handler.coffee:1]: Part of this function looks very duplicated with redmine-handler, can't it be factorized?

File: scripts/gerrit-handler.coffee [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+
+      ret = item(req)
+      if (ret)
+        data = ret
+        return true
+
+    [@dataMethodJSONParse, @dataMethodRaw].forEach(filterChecker)
+
+    # robot.logger.debug "received push: #{util.inspect(data, {depth: null})}"
+    url_subject = data.change_url + "/detail?O=10004"
```
Comment (vthib) [inline scripts/gerrit-handler.coffee]: not used?

File: scripts/gerrit-handler.coffee [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+      return if data
+
+      ret = item(req)
+      if (ret)
+        data = ret
+        return true
+
+    [@dataMethodJSONParse, @dataMethodRaw].forEach(filterChecker)
+
+    # robot.logger.debug "received push: #{util.inspect(data, {depth: null})}"
```
Comment (vthib) [inline scripts/gerrit-handler.coffee]: could be uncommented

File: scripts/gerrit-handler.coffee [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+    url_subject = data.change_url + "/detail?O=10004"
+
+    getUserLogin = (user, finished) ->
```
Comment (vthib) [inline scripts/gerrit-handler.coffee]: should be defined as a real function, not an anonymous one in this function

File: scripts/gerrit-handler.coffee [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+        if err
+          robot.logger.debug err
+        finished stdout.trim()
+
+    getUserLogin(data.change_owner, ((login) ->
+        data.nickname = login
+        getUserLogin(data.author, ((login) ->
+          data.emitter = login
+          robot.emit data.type, data
+          )
```
Comment (vthib) [inline scripts/gerrit-handler.coffee]: weird alignment, should be '))' aligned with second getUserLogin

File: scripts/gerrit-handler.coffee [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+        )
+      )
```
Comment (vthib) [inline scripts/gerrit-handler.coffee]: idem

File: scripts/user-handler.coffee [hunk 10]
```diff
@@ -37,47 +76,20 @@ module.exports = (robot) ->
...
-      msg  = "You are already subscribed to redmine notifications. "
-      msg += "What more could you want :upside_down_face:?\n"
-      msg += "To unsubscribe, use the `redmine unsubscribe` command"
-    else
-      msg  = "You are not subscribed to redmine notifications. "
-      msg += "You're missing out! :sunglasses:\n"
-      msg += "To subscribe, use the `redmine subscribe` command"
-
-    res.send msg
+    on_status(redmine, robot, res)
```
Comment (vthib) [inline scripts/user-handler.coffee]: I don't understand why _redmine_ is not quoted, ti should be afaict

File: scripts/user-handler.coffee [hunk 11]
```diff
@@ -37,47 +76,20 @@ module.exports = (robot) ->
...
-        res.send "Unknown #{cmd} command."
-        return
+    cmd = res.match[1]
+    nickname = res.envelope.user.name
+    on_action(cmd, nickname, 'redmine', robot, res)
 
-      robot.brain.set(key, subs)
+  # Gerrit status
+  robot.respond /gerrit$/, (res) ->
+    on_status(gerrit, robot, res)
```
Comment (vthib) [inline scripts/user-handler.coffee]: idem

**Review by vthib — CHANGES_REQUESTED** [review]

File: scripts/event-handler.coffee (hunk 1 above)
Comment (yannKagan) [reply scripts/event-handler.coffee]: done

File: scripts/gerrit-handler.coffee (hunk 4 above)
Comment (yannKagan) [reply scripts/gerrit-handler.coffee:1]: GerritNotifier and RedmineNotifier now are extended from NotifHandler class

**Comment by yannKagan** [comment]
Done

**Review by vthib — APPROVED** [review]

**Pull request merged**